R/utils.R

```diff
@@ -126,3 +126,73 @@ stop_or_reject <- function(message, async = FALSE) {
   }
   stop(err)
 }
+
+as_predicate <- function(arg, env = rlang::caller_env()) {
+  if(rlang::is_formula(arg) || is.function(arg)) {
+    fun <- rlang::as_function(arg, env = env)
+  } else {
+    fun <- function(x) identical(x, arg)
+  }
+
+  function(...) {
+    res <- fun(...)
+    if(!rlang::is_true(res) && !rlang::is_false(res)) {
+      stop("Predicate functions must return a single `TRUE` or `FALSE`.")
+    }
+    res
+  }
+}
+
+#' Combine predicates
+#'
+#' @param list_of_predicates A named list of predicates.
+#'
+#' @return A function that take a single parameter. The argument of the
+#'     returned function is expected to be a named list. The predicates
+#'     function are applied to the objects of the result
+#' @noRd
+combine_predicates <- function(list_of_predicates) {
+  if(length(list_of_predicates) == 0) return(function(...) TRUE)
+  function(result) {
+    # if a name of a predicate is missing in the result object, return FALSE early
+    if(length(setdiff(names(list_of_predicates), names(result))) > 0) {
+      return(FALSE)
+    }
+    bool <- purrr::imap_lgl(list_of_predicates, ~ .x(result[[.y]]))
+    all(bool)
+  }
+}
+
+
+# callbacks wrappers ------------------------------------------------------
+dewrap <- function(x, ...) {
+  UseMethod("dewrap", x)
+}
+
+dewrap.default <- function(x, ...) {
+  x
+}
+
+dewrap.crrri_callback_wrapper <- function(x, ...) {
+  attr(x, "callback", exact = TRUE)
+}
+
+format.crrri_callback_wrapper <- function(x, ...) {
+  format_object <- paste(collapse = "\n", format(dewrap(x)))
+  paste("=== wrapper over function ===", format_object, sep = "\n")
+}
+
+print.crrri_callback_wrapper <- function(x, ...) {
+  cat(format(x), "\n")
+}
+
+new_callback_wrapper <- function(wrapper_fn, callback) {
+  check_class <- function(x) inherits(x, c("function", "rlang_lambda_function"))
+  stopifnot(check_class(wrapper_fn), check_class(callback))
+  attr(wrapper_fn, "callback") <- dewrap(callback)
+  if(!inherits(wrapper_fn, "crrri_callback_wrapper")) {
+    class(wrapper_fn) <- c("crrri_callback_wrapper", class(wrapper_fn))
```

Comment:
If we were rigorous, I think we should opt for a double inheritance mechanism. But double inheritance in R freaks me out :fearful: 
Saying that a `callback_wrapper` is a subclass of `function` seems conservative to me and less prone to errors.

Reply:
in `rlang`, result of `as_function` is just of class `rlang_lambda_function`. Not function. 
a function is not defined by its class but by its type I think.
Is there a problem to just say that this wrapper_fn is just of class `crrri_callback_wrapper` and is also a function ?

Reply:
I see the point. But my fear is the following:

``` r
wrapper <- rlang::as_function(~ .x)
rlang::is_function(wrapper)
#> [1] TRUE
rlang::is_lambda(wrapper)
#> [1] TRUE
attr(wrapper, "class")
#> [1] "rlang_lambda_function"

# change the class of the dummy fun
class(wrapper) <- "crrri_callback_wrapper"
rlang::is_function(wrapper)
#> [1] TRUE
rlang::is_lambda(wrapper)
#> [1] FALSE
attr(wrapper, "class")
#> [1] "crrri_callback_wrapper"
```

<sup>Created on 2019-05-04 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

Reply:
yes `is_lambda` is a special test for `rlang::as_function` results when a formula is provided. If we never test this or don't use it, it does not matter.

We leave that as is for now and see later. It was just a question, because I am not sure we need to keep everything for what we want to do.

Reply:
Sure, maybe we do not to keep everything and I adopted here a rather conservative approach.